engine/crates/composition/src/emit_federated_graph.rs

```diff
@@ -121,80 +114,101 @@ fn emit_fields<'a>(ir_fields: Vec<FieldIr>, ctx: &mut Context<'a>) {
+            field_type,
+            arguments,
+            resolvable_in,
+            provides,
+            requires,
+            composed_directives,
+            overrides,
+            description,
+        } in fields.drain(..)
+        {
+            let r#type = ctx.insert_field_type(ctx.subgraphs.walk(field_type));
+            let field_name = ctx.insert_string(ctx.subgraphs.walk(field_name));
+
+            let field = federated::Field {
+                name: field_name,
+                r#type,
+                arguments,
+                overrides,
+
+                provides: Vec::new(),
+                requires: Vec::new(),
+                resolvable_in,
+                composed_directives,
+                description,
             };
 
-        match parent_definition {
-            parent @ federated::Definition::Object(object_id) => {
-                let field_id = push_field(ctx, parent, composed_directives);
-                ctx.push_object_field(object_id, field_id);
+            let field_id = federated::FieldId(ctx.out.fields.push_return_idx(field));
+
+            start_field_id = start_field_id.or(Some(field_id));
+            end_field_id = Some(field_id);
+
+            for (subgraph_id, definition, provides) in provides.iter().filter_map(|field_id| {
+                let field = ctx.subgraphs.walk_field(*field_id);
+                field.directives().provides().map(|provides| {
+                    let field_type_name = ctx.insert_string(field.r#type().type_name());
+                    (
+                        federated::SubgraphId(field.parent_definition().subgraph_id().idx()),
+                        ctx.definitions[&field_type_name],
+                        provides,
+                    )
+                })
+            }) {
+                field_provides.push((field_id, subgraph_id, definition, provides));
             }
-            parent @ federated::Definition::Interface(interface_id) => {
-                let field_id = push_field(ctx, parent, composed_directives);
-                ctx.push_interface_field(interface_id, field_id);
+
+            for (subgraph_id, provides) in requires.iter().filter_map(|field_id| {
+                let field = ctx.subgraphs.walk_field(*field_id);
+                field.directives().requires().map(|provides| {
+                    (
+                        federated::SubgraphId(field.parent_definition().subgraph_id().idx()),
+                        provides,
+                    )
+                })
+            }) {
+                field_requires.push((field_id, subgraph_id, definition, provides));
+            }
+
+            let selection_map_key = (definition, field_name);
+            ctx.selection_map.insert(selection_map_key, field_id);
+        }
+
+        let fields = start_field_id
+            .zip(end_field_id)
+            .map(|(start, end)| federated::Fields {
+                start,
+                end: federated::FieldId(end.0 + 1),
```

Comment:
I'm wondering if there isn't a nicer way, this feels ad hoc. I thought about RangeInclusive but that's less familiar.